File: src/enemies/ChainAndBall.cpp
/*
 * Copyright (C) 2009 Christopho, Solarus - http://www.solarus-engine.org
 *
 * Solarus is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Solarus is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
#include "enemies/ChainAndBall.h"
#include "movements/CircleMovement.h"
#include "Sprite.h"
#include "Map.h"

const int ChainAndBall::nb_links;

/**
 * @brief Constructor.
 * @param params name and position of the enemy to create
 */
ChainAndBall::ChainAndBall(const ConstructionParameters &params):
  Enemy(params), center_entity(NULL) {

}

/**
 * @brief Destructor.
 */
ChainAndBall::~ChainAndBall(void) {

}

/**
 * @brief Initializes this enemy.
 */
void ChainAndBall::initialize(void) {

  // attack/defense features
  set_damage(2);
  set_life(1);

  // sprites
  create_sprite("enemies/chain_and_ball");
  get_sprite()->set_current_animation("walking");
  link_sprite = new Sprite("enemies/chain_and_ball");
  link_sprite->set_current_animation("chain");

  // rectangle
  set_size(16, 16);
  set_origin(8, 8);

  // reactions to attacks
  set_no_attack_consequences();
}

/**
 * @brief Updates this enemy.
 */
void ChainAndBall::update(void) {
  Enemy::update();
}

/**
 * @brief Displays the chain and ball on the map.
 */
void ChainAndBall::display_on_map(void) {

  Enemy::display_on_map();

  if (is_visible()) {
    // also display the chain
    for (int i = 0; i < nb_links; i++) {
      map->display_sprite(link_sprite, link_xy[i].get_x(), link_xy[i].get_y());
    }
  }
}

/**
 * @brief Returns whether this entity has to be displayed in y order.
 *
 * This function returns whether the entity should be displayed above
 * the hero and other entities having this property when it is in front of them.
 *
 * @return true if this type of entity is displayed at the same level as the hero
 */
bool ChainAndBall::is_displayed_in_y_order(void) {
  return false; // unlike usual enemies
}

/**
 * @brief Makes the chain and ball follow the specified entity.
 * @param entity the entity to follow
 * @param x x coordinate of where the chain movement should be centered on (relative to the entity followed)
 * @param y y coordinate of where the chain movement should be centered on (relative to the entity followed)
 * @param radius radius of the circles to make, in pixels (see CircleMovement::set_radius)
 * @param radius_speed speed of the radius variation in pixels per second (see CircleMovement::set_radius_speed)
 * @param max_rotations number of rotations to make (see CircleMovement::set_max_rotations)
 * @param loop_delay the movement will restart after this delay in milliseconds (see CircleMovement::set_loop)
 */
void ChainAndBall::attach_to(MapEntity *entity, int x, int y,
    int radius, int radius_speed, int max_rotations, uint32_t loop_delay) {

  clear_movement();

  CircleMovement *movement = new CircleMovement();
  movement->set_center(entity, x, y);
  movement->set_angle_speed(360);
  movement->set_radius_speed(radius_speed);
  movement->set_radius(radius);
  movement->set_max_rotations(max_rotations);
  movement->set_loop(loop_delay);
  movement->start();
  set_movement(movement);

  this->center_entity = entity;
  this->center_entity_dxy = Rectangle(x, y);
}

/**
 * @brief This function is called when the ball has just moved.
 *
 * The chain is then updated.
 */
void ChainAndBall::notify_just_moved(void) {

  Enemy::notify_just_moved();

  if (center_entity != NULL) {
    // recalculate the chain position
    
    int x1 = center_entity->get_x() + center_entity_dxy.get_x();
    int y1 = center_entity->get_y() + center_entity_dxy.get_y();
    int x2 = get_x();
    int y2 = get_y();


    for (int i = 0; i < nb_links; i++) {
      int x = x1 + (x2 - x1) * i / nb_links;
      int y = y1 + (y2 - y1) * i / nb_links;
      link_xy[i].set_xy(x, y);
    }
  }
}

File: quests/zmc/data/enemies/chain_and_ball.lua
local enemy = ...

-- A ball and its chain managed by another enemy.
-- The ball is controlled by a circular movement and
-- the chain is a sprite that automatically fits the space between the other enemy and the ball.
-- They usually disappear when the enemy is killed.

local nb_links = 10
local link_sprite = nil
local link_xy = {
  {x = 0, y = 0},
  {x = 0, y = 0},
  {x = 0, y = 0},
  {x = 0, y = 0},
  {x = 0, y = 0},
  {x = 0, y = 0},
  {x = 0, y = 0},
  {x = 0, y = 0},
  {x = 0, y = 0},
  {x = 0, y = 0}
}
local center_enemy = nil          -- The enemy this chain and ball is attached to.
local center_xy = {x = 0, y = 0}  -- Center point of the circles, relative to the center enemy if any.

function enemy:on_created()

  self:set_life(1)
  self:set_damage(3)
  self:create_sprite("enemies/chain_and_ball")
  self:set_size(16, 16)
  self:set_origin(8, 8)
  self:set_invincible()

  -- Create a second sprite that stays independent of the enemy.
  link_sprite = sol.sprite.create("enemies/chain_and_ball")
  link_sprite:set_animation("chain")

  -- Initialize the links of the chain.
  for i = 1, nb_links do
    link_xy[i].x = 0
    link_xy[i].y = 0
  end
end

function enemy:set_center_enemy(other)

  -- Get the difference of coordinates between me and the center of circles.
  local x, y = self:get_position()
  local other_x, other_y = other:get_position()
  center_xy.x, center_xy.y = x - other_x, y - other_y
  center_enemy = other
end

function enemy:on_pre_draw()

  for i = 1, nb_links do
    self:get_map():draw_sprite(link_sprite, link_xy[i].x, link_xy[i].y)
  end
end

function enemy:on_position_changed(x, y)

  -- Recalculate the chain position.
  if center_enemy ~= nil then
    -- The center is relative to that enemy.
    local x, y = center_enemy:get_position()
    local x1, y1 = x + center_xy.x, y + center_xy.y
    local x2, y2 = self:get_position()

    for i = 1, nb_links do
      link_xy[i].x = x1 + (x2 - x1) * (i - 1) / nb_links
      link_xy[i].y = y1 + (y2 - y1) * (i - 1) / nb_links
    end
  end
end

function enemy:on_restarted()

  local m = sol.movement.create("circle")
  m:set_center(center_enemy, center_xy.x, center_xy.y)
  m:set_radius(56)
  m:set_radius_speed(50)
  m:set_max_rotations(4)
  m:set_loop(2000)
  m:set_angle_speed(360)
  m:start(self)
end

